### PR TITLE
api: add putObject() accepting additional meta data.

### LINKED ISSUE #450: How to set custom metadata to Object as part of putObject call
AWS S3 provides an API to set custom ObjectMetadata using which you can set custom properties such as etag / cache control headers. Is it possible to have ability to set such metadata via MinioClient API?

### PR DESCRIPTION
**Pull request by balamurugana**
Below method is added which facilitates user supplying additional meta data.
```java
public void putObject(String bucketName, String objectName, InputStream stream,
                      long size, Map<String, String> headerMap)
```

fixes #450

### CODE REVIEW

File: api/src/main/java/io/minio/MinioClient.java [hunk 1]
```diff
@@ -3109,6 +3109,77 @@ public void putObject(String bucketName, String objectName, InputStream stream,
...
+   *   builder.append("brave ghost pled; Watch Jeopardy!, Alex Trebeks fun TV quiz game.\n");
+   *   builder.append("---\n");
+   * }
+   * ByteArrayInputStream bais = new ByteArrayInputStream(builder.toString().getBytes("UTF-8"));
+   * // create object
+   * Map<String, String> headerMap = new HashMap<>();
+   * headerMap.put("Content-Type", "application/octet-stream");
+   * minioClient.putObject("my-bucketname", "my-objectname", bais, bais.available(), headerMap);
+   * bais.close();
+   * System.out.println("my-bucketname is uploaded successfully"); }</pre>
```
Comment (NitishT) [inline api/src/main/java/io/minio/MinioClient.java]: s/my-bucketname/my-objectname

File: api/src/main/java/io/minio/MinioClient.java [hunk 2]
```diff
@@ -3109,6 +3109,77 @@ public void putObject(String bucketName, String objectName, InputStream stream,
...
+   * @throws InternalException           upon internal library error
+   *
+   * @see #putObject(String bucketName, String objectName, String fileName)
+   */
+  public void putObject(String bucketName, String objectName, InputStream stream, long size,
+                        Map<String, String> headerMap)
+    throws InvalidBucketNameException, NoSuchAlgorithmException, InsufficientDataException, IOException,
+           InvalidKeyException, NoResponseException, XmlPullParserException, ErrorResponseException,
+           InternalException,
+           InvalidArgumentException, InsufficientDataException {
```
Comment (NitishT) [inline api/src/main/java/io/minio/MinioClient.java]: Looks like an indentation problem here.

File: api/src/main/java/io/minio/MinioClient.java (hunk 1 above)
Comment (balamurugana) [inline api/src/main/java/io/minio/MinioClient.java]: Done

File: api/src/main/java/io/minio/MinioClient.java (hunk 2 above)
Comment (balamurugana) [inline api/src/main/java/io/minio/MinioClient.java]: What is the problem?

Comment (NitishT) [inline api/src/main/java/io/minio/MinioClient.java]: I think both `InvalidArgumentException, InsufficientDataException` can be on the same line as `InternalException`. I have noticed Checkstyle doesn't report such issues, but it doesn't look very good, so I pointed out

Comment (balamurugana) [inline api/src/main/java/io/minio/MinioClient.java]: I see that up to `InternalException` its thrown for all APIs,   After `InternalException`, exceptions listed in new line are specifics to that API.   Its a same thing borrowed from other `putObject()` variants.

Comment (NitishT) [inline api/src/main/java/io/minio/MinioClient.java]: ah..ok, then it's understandable.

**Comment by balamurugana** [comment]
@NitishT No. examples carry commonly used `putObject()` variant i.e. not covering each variant.

**Pull request merged**